### DISCUSSION TITLE
SSR does not pick up environment variables

### DISCUSSION THREAD

**Discussion by bernardoflynn** [vercel/next.js#17261] [Help] [answered]
# Bug report

## Describe the bug

Builds for dev, staging and production contain hard coded environment variables

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Do next build && next export with environment variables NEXT_PUBLIC_CLIENT_BACKEND_API=https://backendurl.com
2. next start and check in browser that url being used by client react app is https://backendurl.com
3. stop the app
4. Change env variable to NEXT_PUBLIC_CLIENT_BACKEND_API=https://backendurl2.com
5. next start and check in browser that url being used by client react app is https://backendurl2.com
6. BUG: browser is still using https://backendurl2.com

## Expected behavior

Backend url should be backendurl2.com as app is doing SSR and not static builds according to Next.js docs on how it decides whether to static or SSR exports per page. Page is dynamic as requires data from external resources and so is not exported as a static html. 

## System information

- OS: MacOS/Linux
- Browser (if applies) [e.g. chrome, safari]
- Version of Next.js: 9.4.4
- Version of Node.js: 14 LTS

**Comment by jamesmosier** [comment, marked as answer]
When you set an environment variable and do a build (or export), Next replaces the `process.env.NEXT_PUBLIC_CLIENT_BACKEND_API` in the actual JavaScript file with your variable. This increases performance by not having to do a lookup for your environment variable each time the code is parsed. 

If you wanted to be able to swap out environment variables and then have them picked up by your server by stopping/starting via `next start` then you could look at [runtime environment variables](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration).

Although as mentioned in that doc:

> Generally you'll want to use build-time environment variables to provide your configuration. The reason for this is that runtime configuration adds rendering / initialization overhead and is incompatible with Automatic Static Optimization.
